ggml/src/ggml-vulkan/ggml-vulkan.cpp

```diff
@@ -2729,15 +3048,29 @@ static void ggml_vk_buffer_memset(vk_buffer& dst, size_t offset, uint32_t c, siz
     dst->device->device.resetFences({ dst->device->fence });
 }
 
-static uint32_t ggml_vk_guess_split_k(int m, int n, int k) {
+static uint32_t ggml_vk_guess_split_k(ggml_backend_vk_context * ctx, int m, int n, int k, const vk_pipeline& pipeline) {
     VK_LOG_DEBUG("ggml_vk_guess_split_k(" << m << ", " << n << ", " << k << ")");
     // if (k > 128 && (m < 128 || n < 128) && m > 2 && n > 2) {
     //     return 4;
     // }
 
-    return 1;
+    uint32_t split_k = 1;
+    // heuristic for coopmat2
+    if (ctx->device->coopmat2 && m >= (int)pipeline->wg_denoms[0] && n >= (int)pipeline->wg_denoms[1]) {
+        // If k is 'large' and the SMs will fill less than halfway, use split_k.
+        uint32_t m_tiles = CEIL_DIV(m, pipeline->wg_denoms[0]);
+        uint32_t n_tiles = CEIL_DIV(n, pipeline->wg_denoms[1]);
+        if (k >= 2048 && m_tiles * n_tiles < ctx->device->shaderSMCount / 2) {
+            split_k = ctx->device->shaderSMCount / (m_tiles * n_tiles);
+            // Clamp to 2 or 4
+            split_k = std::min(split_k, 4u);
+            if (split_k == 3) {
+                split_k = 2;
+            }
+        }
+    }
```

Comment:
Assuming split_k works reliably here, I guess the logic should also work for the non-cm2 codepath. I don't quite remember what issues specifically I had with it to make me disable it. I guess I should look into it again after this change is merged.

Reply:
There was a bug where split_k > 4 didn't work, the fix is at line 3338. Otherwise I didn't run into any functional issues. Whether it helps performance is still very system-dependent and we'd need to do testing to enable it more broadly.